ddtrace/contrib/botocore/patch.py

```diff
@@ -52,23 +78,154 @@ def inject_trace_data_to_message_attributes(trace_data, entry):
+    for entry in entries:
         inject_trace_data_to_message_attributes(trace_data, entry)
 
 
-def inject_trace_to_sqs_message(args, span):
+def inject_trace_to_sqs_or_sns_message(args, span):
+    # type: (List[Any], Span) -> None
+    """
+    :args: contains args for the current botocore action, SNS/SQS record is at index 1
+    :span: the span which provides the trace context to be propagated
+
+    Inject trace headers into MessageAttributes for the SQS or SNS record
+    """
     trace_data = {}
     HTTPPropagator.inject(span.context, trace_data)
     params = args[1]
 
     inject_trace_data_to_message_attributes(trace_data, params)
 
 
+def inject_trace_to_eventbridge_detail(args, span):
+    # type: (List[Any], Span) -> None
+    """
+    :args: contains args for the current botocore action, EventBridge records are at index 1
+    :span: the span which provides the trace context to be propagated
+
+    Inject trace headers into the EventBridge record if the record's Detail object contains a JSON string
+    Max size per event is 256KB (https://docs.aws.amazon.com/eventbridge/latest/userguide/eb-putevent-size.html)
+    """
+    params = args[1]
+    if "Entries" not in params:
+        log.debug("Unable to inject context. The Event Bridge event had no Entries.")
+        return
+
+    for entry in params["Entries"]:
+        detail = {}
+        if "Detail" in entry:
+            try:
+                detail = json.loads(entry["Detail"])
+            except ValueError:
+                log.debug("Detail is not a valid JSON string")
+                continue
+
+        detail["_datadog"] = {}
+        HTTPPropagator.inject(span.context, detail["_datadog"])
+        detail_json = json.dumps(detail)
+
+        # check if detail size will exceed max size with headers
+        detail_size = len(detail_json)
+        if detail_size >= MAX_EVENTBRIDGE_DETAIL_SIZE:
+            log.warning("Detail with trace injection (%s) exceeds limit (%s)", detail_size, MAX_EVENTBRIDGE_DETAIL_SIZE)
+            continue
+
+        entry["Detail"] = detail_json
+
+
+def get_kinesis_data_object(data):
+    # type: (str, Optional[bool]) -> Optional[Dict[str, Any]]
+    """
+    :data: the data from a kinesis stream
+    :try_b64: whether we should try to decode the string as base64
+
+    The data from a kinesis stream comes as a string (could be json, base64 encoded, etc.)
+    We support injecting our trace context in the following two cases:
+    - json string
+    - base64 encoded json string
+    If it's neither of these, then we leave the message as it is.
+    """
+
+    # check if data is a json string
+    try:
+        return json.loads(data)
+    except ValueError:
+        pass
+
+    # check if data is a base64 encoded json string
+    try:
+        return json.loads(base64.b64decode(data).decode("ascii"))
+    except ValueError:
+        raise TraceInjectionDecodingError("Unable to parse kinesis streams data string")
```

Comment:
@hghotra To be clear, this will mean that if the kinesis stream data is not json, this warning will be logged. This is not how data decoding is handled in the other inject calls. Is that correct?

Reply:
Prior to this PR, a mixture or `debug` and `warning` was used to notify the use when a trace injection failed. I’ve updated all `debug` statements to `warning`. it makes sense to show a warning message with the reason for not being able to inject the trace headers. If the user has `distributed_tracing` enabled, they would expect to be notified if a trace injection failed.